RFS/src/main/java/org/opensearch/migrations/bulkload/common/OpenSearchClientFactory.java

```diff
@@ -40,24 +39,23 @@ public OpenSearchClientFactory(ConnectionContext connectionContext) {
     }
 
     public OpenSearchClient determineVersionAndCreate() {
-        if (version == null) {
-            version = getClusterVersion();
-        }
-        var clientClass = getOpenSearchClientClass(version);
-        try {
-            return clientClass.getConstructor(ConnectionContext.class, Version.class)
-                    .newInstance(connectionContext, version);
-        } catch (NoSuchMethodException | InstantiationException | IllegalAccessException | InvocationTargetException e) {
-            throw new ClientInstantiationException("Failed to instantiate OpenSearchClient", e);
-        }
+        return determineVersionAndCreate(null, null);
     }
 
     public OpenSearchClient determineVersionAndCreate(RestClient restClient, FailedRequestsLogger failedRequestsLogger) {
         if (version == null) {
             version = getClusterVersion();
         }
+        if (connectionContext.getCompressionSupported() == null) {
+            var compressionSupported = getCompressionEnabled();
+            connectionContext.setCompressionSupported(compressionSupported);
```

Comment:
Lets avoid making connection context stateful / mutable during the runtime lifecycle.

Reply:
Thanks for making this update